### PR TITLE
cmd/snap-confine: add snap-confine command line parser module

### PR DESCRIPTION
**Pull request by zyga**
This branch adds a dedicated, tested module for snap-confine argument parser. It will soon replace the ad-hoc parsing code that was done in `main()`

### CODE REVIEW

File: cmd/snap-confine/snap-confine-args.c [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+				if (args->security_tag == NULL) {
+					die("cannot allocate memory for security tag");
+				}
+			} else if (args->executable == NULL) {
+				// The second positional argument becomes the executable name.
+				args->executable = strdup(argv[optind]);
+				if (args->executable == NULL) {
+					die("cannot allocate memory for executable name");
+				}
+				// No more positional arguments are required.
```
Comment (niemeyer) [inline cmd/snap-confine/snap-confine-args.c]: The outcome of this logic is a bit awkward. It accepts `snap-confine --version foo bar` and `snap-confine foo --version bar` but not `snap-confine foo bar --version`, with the latter actually handing off the option into the confined command. It seems more clear to be posix-strict here: flags first, positional arguments later. No flags accepted after we observe the first positional argument.

File: cmd/snap-confine/snap-confine-args.c [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+{
+	struct sc_args *args = NULL;
+	struct sc_error *err = NULL;
+
+	if (argcp == NULL || argvp == NULL) {
+		err = sc_error_init(SC_ARGS_DOMAIN, 0,
+				    "cannot parse arguments, argc or argv are NULL");
+		goto out;
+	}
+	// Use local copies of argc and argv.
```
Comment (niemeyer) [inline cmd/snap-confine/snap-confine-args.c]: This message is misleading. This is a local copy of a pointer to an array. Any changes you make to that array are going straight into the real array, and if you change the pointer you are changing a local variable as usual. There's no "copying".

File: cmd/snap-confine/snap-confine-args.c [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+	// "shift" the argument vector left, except for argv[0], to "consume" the
+	// arguments that were scanned / parsed correctly.
+	int i;
+	for (i = 1; optind + i < argc; ++i) {
+		argv[i] = argv[optind + i];
+	}
+	argv[i] = NULL;
+
+	// Write the updated argc and argv back.
+	*argvp = argv;
```
Comment (niemeyer) [inline cmd/snap-confine/snap-confine-args.c]: Why? That's exactly the same pointer that the code took out. There were no updates to that pointer.

File: cmd/snap-confine/snap-confine-args.c [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+	}
+}
+
+void sc_cleanup_args(struct sc_args **ptr)
+{
+	sc_args_free(*ptr);
+	*ptr = NULL;
+}
+
+bool sc_args_is_version_query(struct sc_args *args)
```
Comment (niemeyer) [inline cmd/snap-confine/snap-confine-args.c]: It's not clear why we need all those helpers. If we cannot allocate arguments, we should return an error right then, instead of going through with an unallocated struct and checking that things are not completely broken on every call site.

File: cmd/snap-confine/snap-confine-args.c (hunk 1 above)
Comment (zyga) [inline cmd/snap-confine/snap-confine-args.c]: Good point, we don't need mixed flags/positionals in any way. I'll simplify the code to handle this.

File: cmd/snap-confine/snap-confine-args.c (hunk 3 above)
Comment (zyga) [inline cmd/snap-confine/snap-confine-args.c]: Ah, blast from the past, it used to be updated :). I'll get rid of this, thanks!

File: cmd/snap-confine/snap-confine-args.c (hunk 4 above)
Comment (zyga) [inline cmd/snap-confine/snap-confine-args.c]: If we cannot allocate memory we do fail instantly with an error message.

Those helpers are there in case someone, later, after various changes and evolution of this code, makes a mistake and manages to call those with NULL pointers. To be extra security paranoid I always check for that.

**Comment by zyga** [comment]
This is meant to replace the hacky parser in #2427

File: cmd/snap-confine/snap-confine-args.c [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+struct sc_args *sc_nonfatal_parse_args(int *argcp, char ***argvp,
+				       struct sc_error **errorp)
+{
+	struct sc_args *args = NULL;
+	struct sc_error *err = NULL;
+
+	if (argcp == NULL || argvp == NULL) {
+		err = sc_error_init(SC_ARGS_DOMAIN, 0,
+				    "cannot parse arguments, argc or argv are NULL");
```
Comment (chipaca) [inline cmd/snap-confine/snap-confine-args.c]: no, `argcp` or `argvp` might be null. `argc` can't ever be null, promise!

Comment (chipaca) [inline cmd/snap-confine/snap-confine-args.c]: in seriousness, this is a minuscule it, but if you need to submit another commit for any reason, maybe change those to `&argc` and `&argv` if you don't want `argcp` and `argvp`.

Comment (chipaca) [inline cmd/snap-confine/snap-confine-args.c]: Also: why `cannot $FOO, $REASON!` instead of `cannot $FOO: $REASON!` in this file?

File: cmd/snap-confine/snap-confine-args.c [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+
+	args = calloc(1, sizeof *args);
+	if (args == NULL) {
+		die("cannot allocate memory for command line arguments object");
+	}
+	// Check if we're being called through the ubuntu-core-launcher symlink.
+	// When this happens we want to skip the first positional argument as it is
+	// the security tag repeated (legacy).
+	bool ignore_first_tag = false;
+	char *basename = strrchr(argv[0], '/');
```
Comment (chipaca) [inline cmd/snap-confine/snap-confine-args.c]: So all it takes is to

```C
    execl("snap-confine", NULL);
```

to segfault this? *Interesting...*

File: cmd/snap-confine/snap-confine-args.c [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+	if (basename != NULL) {
+		// NOTE: this is safe because we, at most, may move to the NUL byte
+		// that compares to an empty string.
+		basename += 1;
+		if (strcmp(basename, "ubuntu-core-launcher") == 0) {
```
Comment (chipaca) [inline cmd/snap-confine/snap-confine-args.c]: may I interest you in `streq` (in `system-shutdown-utils`)?

File: cmd/snap-confine/snap-confine-args.c [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+					    argv[optind]);
+			goto out;
+		}
+	}
+
+	// Parse positional arguments.
+	//
+	// NOTE: optind is not reset, we just continue from where we left off in
+	// the loop above.
+	for (; optind < argc; ++optind) {
```
Comment (chipaca) [inline cmd/snap-confine/snap-confine-args.c]: why is this a loop at all?

File: cmd/snap-confine/snap-confine-args.c [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+	if (args->executable == NULL) {
+		err = sc_error_init(SC_ARGS_DOMAIN, SC_ARGS_ERR_USAGE,
+				    "executable name was not provided");
+		goto out;
+	}
+
+	int i;
+ done:
+	// "shift" the argument vector left, except for argv[0], to "consume" the
+	// arguments that were scanned / parsed correctly.
```
Comment (chipaca) [inline cmd/snap-confine/snap-confine-args.c]: Why not move argv itself?

**Comment by chipaca** [comment]
to be clear, the "changes requested" is about the use of `strrchr` on an unchecked pointer. The rest are nits.

File: cmd/snap-confine/snap-confine-args.c [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+	char *executable;
+
+	// Flag indicating that --version was passed on command line.
+	bool is_version_query;
+	// Flag indicating that --classic was passed on command line.
+	bool is_classic_confinement;
+};
+
+struct sc_args *sc_nonfatal_parse_args(int *argcp, char ***argvp,
+				       struct sc_error **errorp)
```
Comment (stolowski) [inline cmd/snap-confine/snap-confine-args.c]: Would it make more sense to make the function return sc_error (and return sc_args by value-argument of the function) instead and mark the function with __attribute__ ((warn_unused_result)) to force the caller to do something about the error?

Comment (zyga) [inline cmd/snap-confine/snap-confine-args.c]: No, because it is still valid to pass a NULL `errorp` here and anywhere else (the behavior is predictable and correct) and this might actually be more convenient. Otherwise if you actually want to die you force everyone to collect the error, check it and then die. While in this specific function I don't think this is useful, as a general theme it feels wrong.

File: cmd/snap-confine/snap-confine-args.c (hunk 9 above)
Comment (zyga) [inline cmd/snap-confine/snap-confine-args.c]: Because of argv[0]. Also because this is what various other parsers do and (maybe some black magic) I wanted to stay in line.

File: cmd/snap-confine/snap-confine-args.c (hunk 7 above)
Comment (zyga) [inline cmd/snap-confine/snap-confine-args.c]: Yeah, I would love to add that to to `string-utils.[ch]`. Note that this branch pre-dates your pull request. I will gladly refactor it to simplify this later.

File: cmd/snap-confine/snap-confine-args.c (hunk 6 above)
Comment (zyga) [inline cmd/snap-confine/snap-confine-args.c]: Are you sure? I think I check for argc size above. https://github.com/snapcore/snapd/pull/2416/files/19c093cfbbb28775ef98801038fb89092d1dedb4#diff-5c0eca39cb68ec44020b045836523bbaR51

File: cmd/snap-confine/snap-confine-args.c (hunk 5 above)
Comment (zyga) [inline cmd/snap-confine/snap-confine-args.c]: I'll answer the last comment first, no reason, I need to review what the common way is and just apply that everywhere.

Comment (zyga) [inline cmd/snap-confine/snap-confine-args.c]: As for `argc` and `argv` vs `argcp` and `argvp` -- good catch! I'll correct that.

File: cmd/snap-confine/snap-confine-args.c (hunk 8 above)
Comment (zyga) [inline cmd/snap-confine/snap-confine-args.c]: Because we collect positional arguments here. Am I missing anything?

Comment (chipaca) [inline cmd/snap-confine/snap-confine-args.c]: How is this different from
```C
if (ignore_first_tag) optind++;
args->security_tag = strdup(argv[optind++]);
args->executable = strdup(argv[optind++]);
```
(skipping error checks for clarity). You put yourself in a loop of length two, with an `if (first) {...} else if (second) {...}` as the only body?

Comment (zyga) [inline cmd/snap-confine/snap-confine-args.c]: I see what you mean. Let me think about this.

File: cmd/snap-confine/snap-confine-args.c [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+	char *basename = strrchr(argv[0], '/');
+	if (basename != NULL) {
+		// NOTE: this is safe because we, at most, may move to the NUL byte
+		// that compares to an empty string.
+		basename += 1;
+		if (strcmp(basename, "ubuntu-core-launcher") == 0) {
+			ignore_first_tag = true;
+		}
+	}
+	// Parse option switches.
```
Comment (tyhicks) [inline cmd/snap-confine/snap-confine-args.c]: What's the reason for reimplementing libc's getopt(3)?

Comment (tyhicks) [inline cmd/snap-confine/snap-confine-args.c]: I'm going to hold off on reviewing this loop until I hear back on this.

File: cmd/snap-confine/snap-confine-args.c [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+
+bool sc_args_is_classic_confinement(struct sc_args * args)
+{
+	if (args == NULL) {
+		die("cannot obtain classic confinement flag from NULL argument parser");
+	}
+	return args->is_classic_confinement;
+}
+
+char *sc_args_security_tag(struct sc_args *args)
```
Comment (tyhicks) [inline cmd/snap-confine/snap-confine-args.c]: Do you want callers of this function to be able to modify the security_tag string? I think returning a `const char *` would be cleaner.

File: cmd/snap-confine/snap-confine-args.c [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+{
+	if (args == NULL) {
+		die("cannot obtain security tag from NULL argument parser");
+	}
+	return args->security_tag;
+}
+
+char *sc_args_executable(struct sc_args *args)
```
Comment (tyhicks) [inline cmd/snap-confine/snap-confine-args.c]: Same here regarding `const char *`.

File: cmd/snap-confine/snap-confine-args.h [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+ **/
+bool sc_args_is_classic_confinement(struct sc_args *args);
+
+/**
+ * Get the security tag passed to snap-confine.
+ *
+ * The return value may be NULL if snap-confine was invoked with --version. It
+ * is never NULL otherwise.
+ *
+ * The return value doesn't need to be freed(). It is bound to the lifetime of
```
Comment (tyhicks) [inline cmd/snap-confine/snap-confine-args.h]: s/doesn't need to be freed()/must not be freed()/

File: cmd/snap-confine/snap-confine-args.h [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+ **/
+char *sc_args_security_tag(struct sc_args *args);
+
+/**
+ * Get the executable name passed to snap-confine.
+ *
+ * The return value may be NULL if snap-confine was invoked with --version. It
+ * is never NULL otherwise.
+ *
+ * The return value doesn't need to be freed(). It is bound to the lifetime of
```
Comment (tyhicks) [inline cmd/snap-confine/snap-confine-args.h]: s/doesn't need to be freed()/must not be freed()/

File: cmd/snap-confine/snap-confine-args.h (hunk 14 above)
Comment (tyhicks) [inline cmd/snap-confine/snap-confine-args.h]: I say "must not" because if some future caller frees the pointer, then sc_args_free() will trigger a double-free. Might as well be explicit in the function documentation.

Comment (zyga) [inline cmd/snap-confine/snap-confine-args.h]: Yeah, totally agreed. Thanks!

File: cmd/snap-confine/snap-confine-args.c [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+	// Use dereferenced versions of argcp and argvp for convenience.
+	int argc = *argcp;
+	char **const argv = *argvp;
+
+	if (argc == 0 || argv == NULL) {
+		err = sc_error_init(SC_ARGS_DOMAIN, 0,
+				    "cannot parse arguments, argc is zero or argv is NULL");
+		goto out;
+	}
+	// Sanity check, look for NULL argv entries.
```
Comment (zyga) [inline cmd/snap-confine/snap-confine-args.c]: Interesting observation. This was in response to a question from @chipaca (that we don't check for NULL arguments). I think you are right. Since we don't (and cannot really) check argc in a reasonable way, checking argv is equally unreasonable.

I was looking if there's a way to attack snap-confine by exec'ing it with a malicious argument vector. Suggestions are very much welcome.

File: cmd/snap-confine/snap-confine-args.c (hunk 11 above)
Comment (zyga) [inline cmd/snap-confine/snap-confine-args.c]: Mostly flexibility. All the various command line parsers have their quirks and we are often asked to have precise semantics. It also lets us be sure that there's nothing we're missing as to how this simple but critical part behaves.

File: cmd/snap-confine/snap-confine-args.c (hunk 16 above)
Comment (tyhicks) [inline cmd/snap-confine/snap-confine-args.c]: I think ensuring that argc meets a minimum value (whatever the appropriate amount is for progname and minimum number of required args) is as good of a check as you're going to get.

File: cmd/snap-confine/snap-confine-args.c (hunk 11 above)
Comment (tyhicks) [inline cmd/snap-confine/snap-confine-args.c]: Ok. I think libc's built in parser is likely to be secure and I find it simple to use but I can appreciate the desire to shed some of the unneeded features that it has.
